### PR TITLE
Implemented NS&I connection health test

### PR DESCRIPTION
**Pull request by AndyHWChung**

### CODE REVIEW

File: app/uk/gov/hmrc/helptosavefrontend/config/FrontendAppConfig.scala [hunk 1]
```diff
@@ -79,6 +79,8 @@ object FrontendAppConfig extends AppConfig with ServicesConfig {
 
   val nsiUpdateEmailUrl: String = s"${baseUrl("nsi")}${getString("microservice.services.nsi.update-email-url")}"
 
+  val nsiTestUrl: String = s"${baseUrl("nsi")}${getString("microservice.services.nsi.test-url")}"
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/config/FrontendAppConfig.scala]: can we rename the `test-url` as `helath-check-url` ?

File: app/uk/gov/hmrc/helptosavefrontend/config/HealthCheckModule.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+package uk.gov.hmrc.helptosavefrontend.config
+
+import com.google.inject.AbstractModule
+import uk.gov.hmrc.helptosavefrontend.health.NSIConnectionHealthTest
+
+class HealthCheckModule extends AbstractModule {
+
+  override def configure() = {
+    bind(classOf[NSIConnectionHealthTest])
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/config/HealthCheckModule.scala]: can we rename `NSIConnectionHealthTest`  to something like `NSIConnectionHealthCheck`
`Test` sounds like a junit test :)

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 3]
```diff
@@ -46,6 +45,8 @@ trait NSIConnector {
 
   def updateEmail(userInfo: NSIUserInfo)(implicit hc: HeaderCarrier, ex: ExecutionContext): Result[Unit]
 
+  def test(userInfo: NSIUserInfo)(implicit hc: HeaderCarrier, ex: ExecutionContext): Result[Unit]
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: `healthCheck`

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 4]
```diff
@@ -130,6 +131,18 @@ class NSIConnectorImpl @Inject() (conf: Configuration, metrics: Metrics) extends
       }
   }
 
+  override def test(userInfo: NSIUserInfo)(implicit hc: HeaderCarrier, ex: ExecutionContext): Result[Unit] = EitherT[Future, String, Unit]{
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: i guess we need to add the implicit headers as well `(nsiUserInfoFormat, hc.copy(authorization = None), ec)` based on my recent changes or rebase your branch to latest master and see

File: app/uk/gov/hmrc/helptosavefrontend/health/NSIConnectionHealthCheck.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import uk.gov.hmrc.helptosavefrontend.models.NSIUserInfo.ContactDetails
+import uk.gov.hmrc.helptosavefrontend.util.Email
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.Future
+import scala.util.control.NonFatal
+
+class NSIConnectionHealthCheck @Inject()(system: ActorSystem, configuration: Configuration, metrics: Metrics, nSIConnector: NSIConnector) {
+
+  val healthCheck: ActorRef = system.actorOf(
```
Comment (sureshhmrc) [inline app/uk/gov/hmrc/helptosavefrontend/health/NSIConnectionHealthCheck.scala]: is this val `healthCheck `  being referenced anywhere ?

File: app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+ *
+ * @param name The name of the test
+ * @param config The config variables will be looked for under the namespace `health-check.{name}` where `name` is
+ *               the name of the test
+ * @param scheduler      The scheduler used to schedule the health tests
+ * @param metrics        The metrics to be updated
+ * @param pagerDutyAlert A function which will alert pager duty
+ * @param runnerProps    The props of the tests to be run
+ */
+class HealthCheck(name:             String,
```
[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: I intended `HealthCheck` to be generic - I wanted it to work for any test we may have in the future

File: app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                  config:           Config,
+                  scheduler:        Scheduler,
+                  metrics:          Metrics,
```
[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: to implement the graph in grafana where we see the number of test failures we need the metrics

File: app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    config.get[FiniteDuration](s"health.$name.poll-timeout").value
+
+  val numberOfTestsBetweenUpdates: Int =
+    config.get[Int](s"health.$name.poll-count-between-updates").value
+
+  val maximumConsecutiveFailures: Int =
+    config.get[Int](s"health.$name.poll-count-failures-to-alert").value
+
+  val numberOfTestsBetweenAlerts: Int =
+    config.get[Int](s"health.$name.poll-count-between-pager-duty-alerts").value
```
[the remark this answers is not in the document]
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: I used `config.get[Int]` to be consistent with `config.get[FiniteDuration]`

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 9]
```diff
@@ -130,6 +131,18 @@ class NSIConnectorImpl @Inject() (conf: Configuration, metrics: Metrics) extends
       }
   }
 
+  override def healthCheck(userInfo: NSIUserInfo)(implicit hc: HeaderCarrier, ex: ExecutionContext): Result[Unit] = EitherT[Future, String, Unit]{
+    httpProxy.put(nsiHealthCheckUrl, userInfo, Map(nsiAuthHeaderKey → nsiBasicAuth))(nsiUserInfoFormat, hc.copy(authorization = None), ex)
+      .map[Either[String, Unit]] { response ⇒
+        response.status match {
+          case Status.OK ⇒ Right(())
+          case other     ⇒ Left(s"Received unexpected status $other from NS&I while trying to update email. Body was ${response.body}")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: Is this text correct? Maybe...
HealthCheck received an unexpected http status code $other from NS&I. Body was blah, blah

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala [hunk 10]
```diff
@@ -130,6 +131,18 @@ class NSIConnectorImpl @Inject() (conf: Configuration, metrics: Metrics) extends
...
+        }
+      }.recover {
+        case e ⇒ Left(s"Encountered error while trying to create account: ${e.getMessage}")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: Is this text correct? Maybe...
HealthCheck encountered an error while trying to communicate with NS&I: ${e.getMessage}

File: app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala (hunk 9 above)
Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/connectors/NSIConnector.scala]: That `String` will eventually be picked up and logged in this line here in `HeathCheck`:

https://github.com/hmrc/help-to-save-frontend/blob/master/app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala#L110

File: app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (tim-holbrook) [reply app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: Ah, if this is the case, then the words "NS&I" need to be removed from all comments and replaced with "endpoint" or suchlike

File: app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+
+  var currentRunnerProps: Props = propsQueue.next()
+
+  var performTestTask: Option[Cancellable] = None
+
+  override def receive: Receive = ok(0)
+
+  def ok(count: Int): Receive = performTest orElse {
+    case HealthCheckResult.Success(nanos) ⇒
+      log.info(s"Health check $name OK ${timeString(nanos)}")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: Can we start all comments wth say 'HealthCheck' allowing us to search for them in Kibana by ^HealthCheck

Comment (AndyHWChung) [reply app/uk/gov/hmrc/helptosavefrontend/health/HealthCheck.scala]: yeah, that makes sense

File: app/uk/gov/hmrc/helptosavefrontend/health/NSIConnectionHealthCheck.scala [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+      () ⇒ (),
+      () ⇒ (),
+      NSIConnectionHealthCheckRunner.props(nSIConnector, metrics, Payload.Payload1),
+      NSIConnectionHealthCheckRunner.props(nSIConnector, metrics, Payload.Payload2)
+    )
+  )
+
+  // start the health check only if it is enabled
+  if (enabled) {
+    logger.info(s"$name health test enabled - starting HealthCheck actor")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/health/NSIConnectionHealthCheck.scala]: How about...
"HealthCheck for $name is enabled, commencing connection checking period timer every XXXs"
...plus can we change the word 'test' or 'Test' in the code to check

File: app/uk/gov/hmrc/helptosavefrontend/metrics/Metrics.scala [hunk 13]
```diff
@@ -72,6 +74,10 @@ class Metrics @Inject() (metrics: com.kenshoo.play.metrics.Metrics) {
 
   val authTimer: Timer = timer("frontend.auth-timer")
 
+  val healthCheckFailuresHistogram: Histogram = histogram("frontend.health-check.nsi.failures")
+
+  val healthCheckTimer: Timer = timer("frontend.health-check.nsi.timer")
```
Comment (tim-holbrook) [inline app/uk/gov/hmrc/helptosavefrontend/metrics/Metrics.scala]: frontend.health-check  ...or ...  frontend.nsi-account-creation-time  ?
Additionally, pls can these prop names be as outlined in the Jira ticket

File: conf/application.conf [hunk 14]
```diff
@@ -346,3 +348,18 @@ mongodb {
 
 # must be a Base64 encoded string of 128/192/256 bits
 crypto.encryption-key = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA="
+
+
+health {
+
+  nsi-connection {
+    enabled = true
+    poll-period = 1 minute
```
Comment (tim-holbrook) [inline conf/application.conf]: Pls set to 3 mins

File: conf/application.conf [hunk 15]
```diff
@@ -346,3 +348,18 @@ mongodb {
...
+    minimum-poll-period = 1 minute
+    poll-timeout = 30 seconds
```
Comment (tim-holbrook) [inline conf/application.conf]: Pls set - following a late Fri meeting - this to 50s, as NS&I's longest timeout has been set to 40s and MDTP is 60s

File: conf/application.conf [hunk 16]
```diff
@@ -346,3 +348,18 @@ mongodb {
...
+    poll-count-between-updates = 20
+    poll-count-failures-to-alert = 3
+    poll-count-between-pager-duty-alerts = 10
```
Comment (tim-holbrook) [inline conf/application.conf]: Pls set to 20 (aka 1hr)

**Pull request merged**